torchtune/models/llava_next/image_processor.py

```diff
@@ -0,0 +1,344 @@
+        )
+
+        # all patches are resized so that smallest side is self.patch_size.
+        # this allows us to crop the images to [self.patch_size, self.patch_size]
+        image_patches = [
+            F.resize(patch, [self.patch_size], interpolation=self.resample).float()
+            for patch in image_patches
+        ]
+
+        # Cropping only happens if the patch is not already a square. For example:
+        # the image is resized and padded to best_resolution 400x300.
+        # Assuming patch size is 200, the generated patches would be of size [200x200, 200x100, 200x200, 200x100]
+        # The patches are then resized to have smallest_side = patch_size.
+        # So the second and last patches would be resized to 400x200, and finally
+        # center_cropped to 200x200. First and third patches are not cropped.
+        image_patches = [self.center_crop(patch) for patch in image_patches]
+
+        # original image is resized, with distortion, to a square of side self.patch_size
+        # and concatenated to patches
+        patch_sized_image = F.resize(
+            image,
+            [self.patch_size, self.patch_size],
+            interpolation=self.resample,
+        ).float()
+
+        image_patches = torch.stack([patch_sized_image] + image_patches)
+
+        return image_patches
+
+
+class ImageProcessor(nn.Module):
+    def __init__(
+        self,
+        patch_size: int = 336,
+        possible_resolutions: Optional[Tuple[Tuple[int, int]]] = None,
+        max_num_chunks: int = 4,
+        resample: str = "bicubic",
+        do_rescale: bool = True,
+        rescale_factor: Union[int, float] = 1 / 255,
+        do_normalize: bool = True,
+        image_mean: Optional[Union[float, Tuple[float]]] = None,
+        image_std: Optional[Union[float, Tuple[float]]] = None,
+    ) -> None:
+        """
+        Args:
+            patch_size (int): Size of the patches to divide the image into.
+            possible_resolutions (Optional[List[Tuple[int, int]]]): List of possible resolutions as tuples (height, width).
+            max_num_chunks (int): Only used possible_resolutions is NOT given. Maximum number of chunks for processing
+            high-resolution images.
+            resample (str): Resampling method used when resizing images.
+            do_rescale (bool): Flag to determine whether to rescale the image by `rescale_factor`.
+            rescale_factor (Union[int, float]): Scale factor used if rescaling the image.
+            do_normalize (bool): Flag to determine whether to normalize the image.
+            image_mean (Optional[Union[float, List[float]]]): Mean values for normalization.
+            image_std (Optional[Union[float, List[float]]]): Standard deviation values for normalization.
+        """
+        super().__init__()
+
+        logger.info("Initializating ImageProcessor...")
+
+        if image_mean is None:
+            image_mean = [0.48145466, 0.4578275, 0.40821073]
```

Comment:
where are these magic numbers from and why do we use these?

Reply:
My confident guess is that this is the mean/std of the dataset used to pretrain the image encoder, and should be used for any subsequent training/inference. Kind like the learned batch normalization parameters.

The reason I dont put it as default is because its bad practice to have lists there, since they are mutable.

It comes from the model config in HF

Reply:
I see, makes sense. Could we place it as a constant at the top of the file and make it a tuple?